### ISSUE TITLE
copy some ideas from vipsthumbnail

### ISSUE THREAD

**Issue by jcupitt** [evadne/supervised-scaler#1]
Hi, I liked your talk. 

You could improve speed and quality quite a bit by stealing some code from `vipsthumbnail`:

https://github.com/jcupitt/libvips/blob/master/tools/vipsthumbnail.c

I think the big things would be exploiting shrink-on-load, colour management, premultiply, and handling rounding errors. Auto-rotate is probably necessary too.

The logic in `vipsthumbnail` should probably be folded into the core library, it's a pain to reimplement every time.

**Comment by evadne** [comment]
Thanks: I shall copy a bit more of that indeed and will leave this issue open until the request has been fulfilled.

**Comment by evadne** [comment]
Hello. I’ve copied most of the ideas but skipped some. Have made a few changes as well.
- Skipped image cropping since that use case has not come up. Also skipped the fit/cover bits so the code will only scale to fit, since the other use case has also not come up. I usually implement scale-to-cover with CSS (`background-image`) when needed so the same image can be used in multiple places.
- Colour management simplified and the program puts out sRGB only since I have not come up against other use cases. May need to put back later.
- Added a bit of code to empty all cached information (see `CACHE_BYTES_ALLOWED_WITHIN_EACH_PASS` and `CACHE_BYTES_ALLOWED_BETWEEN_EACH_PASS`), the goal is to not hold any memory between conversions; it seems that this incurs a hit of ~20ms on my laptop, but the actual hit remains to be tested. May need to be made configurable later.
- Have written the program with C99 in mind so instead of `gboolean` have used `bool`.
- Linear processing disabled but left in as a `#define`. Perhaps this is not correct, will look later. Ideally the program should know when to force that instead of asking for it as an argument.

I’ve managed to have this program run continuously with no leaks:

```
make && yes "1024 1024 /Users/evadne/Desktop/viirs_9apr2015.jpg" | ./priv_dir/convert
```

And I’ve managed to have it deal with a very large image (15000 by 15000, JPEG) with 4% of time required compared to ImageMagick. Previously without shrink-on-load it was taking roughly 33% of time.

You can see the updated C program here — https://github.com/evadne/supervised-scaler/blob/master/deps_local/resampler/c_src/convert.c

**Comment by jcupitt** [comment]
That's a good speedup! Nice work. 

I've started making a `thumbnail` operation in the core library now, so perhaps this will be less work in future. 

https://github.com/jcupitt/libvips/tree/add-vips-thumbnail

You can make imagemagick do shrink-on-load as well. You do something like:

```
$ convert -define jpeg:size=256x256 huge.jpg -resize 128x128 small.jpg
```

That means: load `huge.jpg`, opening in such a way as to get a size of 256x256 or above, resize to 128x128, and save. You need to do at least 200% of the resize with `-resize` and not all with shrink-on-load or you'll see nasty aliasing artifacts. It might be a fairer comparison to your resize operation.

**Comment by jcupitt** [comment]
Oh, the big thing colour management for is handling CMYK images. CMYK profiles can be huge, maybe 500kb, and it's common to not embed them in the source image. 

Colour management can also shrink your thumbnails by x2 or more. If you transform to sRGB (perhaps you have some Adobe98 sources), you can strip all profiles from thumbnails, which for a page full of thumbs can make quite a difference to the amount of data clients need to download.

**Comment by evadne** [comment]
Thanks again,
1.  Indeed it would be fairer if I could get shrink-on-load going in ImageMagick as well, I’ll look into how that can be worked into a dependency I am currently using. Once updated I shall re-run the benchmark.
2. Very good news that a new thumbnail operation was added. I have tried to look into detecting the version of libVIPS involved when confronted with the `centre` option added to `vips_resize` recently.  Given that the library is to be deployed [onto systems with libVIPS already installed in packages](http://packages.ubuntu.com/search?keywords=libvips) presumably maintained by others I will probably have to either wait until a newer package is available or bundle both. Currently my deployment environments still run Ubuntu 14.04 which does not exactly have the latest libraries :(
3. Noted on colour management, will look into whether I have correctly implemented it or if it needs fixing.

**Comment by evadne** [comment]
`time`:

```
$ time vipsthumbnail -s 128 ~/Desktop/viirs_9apr2015.jpg                                                  
> … 0.49s user 0.03s system 98% cpu 0.522 total

$ time convert -define jpeg:size=256x256 ~/Desktop/viirs_9apr2015.jpg -resize 128x128 ~/Desktop/small.jpeg
> … 0.57s user 0.02s system 98% cpu 0.602 total

$ time vipsthumbnail -s 1024 ~/Desktop/viirs_9apr2015.jpg
> … 0.71s user 0.04s system 117% cpu 0.637 total

$ time convert -define jpeg:size=2048x2048 ~/Desktop/viirs_9apr2015.jpg -resize 1024x1024 ~/Desktop/small.png
> … 1.02s user 0.08s system 99% cpu 1.112 total
```

**Comment by jcupitt** [comment]
Yes, exactly, if you are shrinking JPEG files, almost all the work can be done by libjpeg shrink-on-load, the image processing library makes little difference, since it's not processing many pixels. You need to be shrinking images by much more than a factor of 8 (the maximum that libjpeg can do), or you need a different image format. 

TIFF probably shows the largest difference: libtiff is quick and it has no shrink-on-load system that can be exploited, so most CPU time is burnt in the image processing library.

You'd think PNG would be a good test, but libpng is extremely slow, so a good chunk of runtime ends up in there.

**Comment by jcupitt** [comment]
For example:

```
$ time convert wtc.tif -resize 128x128 x.tif
real    0m2.601s
user    0m4.304s
sys 0m0.320s
$ time vipsthumbnail wtc.tif -s 128x128 -o x.tif
real    0m0.290s
user    0m0.184s
sys 0m0.100s
```
